typehero/typehero
# feat: allow solution comments to be shared - Closes #648

Allows sharing `Solution Comments` with the URL format
`http://localhost:3000/challenge/13/solutions/57/comments/175?replyId=176`

## Description


- Added `/challenge/[id]/solutions/[solutionId]/comments/[commentId]/page.tsx` to allow linking to `Solution Comments`

- Moved `<Comments />` outside of `<SolutionDetail />` to be able be used in 
`/challenge/[id]/solutions/[solutionId]/page.tsx` and
`/challenge/[id]/solutions/[solutionId]/comments/[commentId]/page.tsx`

- `<Comments />` will now be required to accept a `challengeId` , previously we were creating `Solution Comments` without having this saved in the DB under `Shared Solutions` data model.
This is because `rootId` can refer to a challenge Id when `<Comment type='CHALLENGE' />` but when [Comment type is 'SOLUTION' rootId refers to `rootSolutionId` and `rootChallengeId` is lost.](https://github.com/typehero/typehero/pull/661/files#diff-6ffcd17d428ccca4e0e17df4f046a75acc8e07b7448591d397466866fac97756L50)
But it is required to create a valid URL.
This could probably be cleaned up.

- created `getPreselectedSolutionCommentMetadata(solution, commentId)` server action in [getCommentRouteData.ts](https://github.com/typehero/typehero/compare/main...zawarudo:typehero:feature/linkable-solution-comments?expand=1#diff-695d0c4e63fb1bf65662842f1fd183848e316140649e9c64e6514b4dd8810108)

- Expanded type `PreselectedCommentMetadata` to allow the type of data from `getPreselectedSolutionCommentMetadata()`
Maybe this is not the best solution, I am still new to TS / Prisma

## Motivation and Context


https://github.com/typehero/typehero/issues/648

## How Has This Been Tested?


- Create a new comment on a submitted solution
- Create another comment replying to the comment above
- Hit share on both comments and visit the links,
- confirm we see the relevant comment with the highlight animation playing.
- Repeat steps under `Description` section of challenge to test comments there.

## Screenshots/Video (if applicable):

Visiting the top comment and its reply from the share function links.
`http://localhost:3000/challenge/13/solutions/57/comments/175`
`http://localhost:3000/challenge/13/solutions/57/comments/175?replyId=176`
![image](https://github.com/typehero/typehero/assets/2955134/cc34845b-2375-48ab-8fe0-ef5bce5f72a8)
